Godot 3.x -> Godot 4.0 project converter

qarmin opened the pull request:
Implementation of project converter(currently 3.x -> 4.0, but it can be easily upgraded to handle new versions) https://github.com/godotengine/godot-proposals/issues/387

~This PR contains #49053, so it would be good to merge that PR before this because it is needed to check if all functions exists even in builtin types.~

Most of things are renamed by "whole word" feature and some with simple Regex.

Before conversion I suggest to use this tool - https://github.com/Scony/godot-gdscript-toolkit to unify code look with some big line length(converter doesn't support multi line expressions and won't support, because it would be too complicated) - e.g. `find ./ | grep "\.gd$" | xargs gdformat --line-length 400`

For me this is the best way to upgrade project:
1. Run tool to validate conversion `godot4 --validation-conversion-3to4 --headless`(current directory must contains godot project)
2. Read output of previous command, validate variables and rename manually names which you don't want to rename automatically by converter
```
Checking for conversion - 23/138 file - "test/resources/test_assert_setget_test_objects/ControlChildMock.tscn" with size - 5 KB
    Checking file took 54 ms.
		Line (7) margin_bottom -> offset_bottom  -  LINE """ margin_bottom = 46.0 """
		Line (15) margin_bottom -> offset_bottom  -  LINE """ margin_bottom = 14.0 """
		Line (6) margin_right -> offset_right  -  LINE """ margin_right = 130.0 """
		Line (24) margin_top -> offset_top  -  LINE """ margin_top = -14.0 """

```
3. Make sure that your game/project works fine in Godot 3
4. Repeat steps from 1, until you will be sure that you want to rename all listed elements
5. Save changes to e.g. Git
6. Run converter `godot4 --convert-3to4 --headless`
7. Try to change project to be able to run it in Godot 4
8. Revert to Godot 3 project
9. Re-apply changes(of course if possible) which have been done in Godot 4, to speedup later iteration of conversion
10. Make sure that your game/project works fine in Godot 3
11. Repeat steps from 6, until there is no more things that can be changed in Godot 3, to simplify conversion and unify codebase with Godot 4 project

Also during converting, this text about current progress is printing to console 
```
Trying to convert	65/138 file - "test/unit/test_stubber.gd" with size - 7 KB
    File was changed, conversion took 131 ms.
Trying to convert	66/138 file - "test/unit/test_utils.gd" with size - 2 KB
    File was not changed, checking took 50 ms.
.....
Converting ended - all files(802), converted files(319), not converted files(483).
```

**Done things:**
- [x] - functions (also in builtin types)
- [x] - classes
- [x] - enums/constants
- [x] - properties
- [x] - tool to validate things which will be changed
- [x] - gdscript keywords
- [x] - signals
- [X] - shaders(only 3 renames, but I can't find more)
- [X] - advanced GDScript renames(like `connect` -> `Callable`) - only some, but can be easily added
- [X] - C# (1:1 CamelCase conversion from GDScript functions)
- [X] - Check if all new functions exists in Godot 4

**Things which may be implemented in future**
- [ ] - Regex caching(Not sure if possible)
- [ ] - Removing code duplication(for now there are some code duplication between validating and converting tool, which proably could be integrated into one function)
- [ ] - csproj conversion

**Problematic things:**
- **Cyclic renames** - e.g. in Skeleton3D `choose_it` change to `select_this`, but Node3D changes `select_this` to `choose_it`. So running tool, will renames all `select_this` functions to `choose_it`(assuming that this records have bigger index in `function_renames` array). Depends of the usage, some things may be renamed(this will broke other functions) when class is used more(like Node3D vs HTTPRequest).
- **Class split** - Some classes like `Geometry` were split at two classes `Geometry2D` and `Geometry3D`, so all instances of `Geometry` are renamed to `Geometry2D`. If proper class is `Geometry3D`, then this needs to be fixed manually
- **Deferred functions** - custom functions renaming(e.g. via regex), works only with proper syntax(arguments, parentheses etc.) so things like `call_deffered` which use little different syntax, won't be always fully renamed.
- **Properties changes** - most properties are named by single or two simple words like `alt`. Renaming them, would broke a lot of users scripts, because this words are usable often.
- **Tweens** - Tweens were completely rewritten(#41794) and now as base they use RefCounted instead Node, so can't be automatically converted. If want to have less amount of problems with converting, don't add Tweens to tree(try to instantitate them where needed). You may also use TweenSequence from  https://github.com/godot-extended-libraries/godot-next/pull/50. 
- **Too big files** - Files with size greater than 5MB are not checked, because it takes too long. If you have such files, consider saving e.g. resources or objects to different file.(Not sure about this, maybe big files will be re-enabled later)
- **await** - all functions which use await keyword must have await word before
- **Windows** - Windows are not longer children of Popups/Controls, so this may require additional changes to code/project

Tested projects:
- [Pole Mrowkowe](https://github.com/qarmin/Pole-Mrowkowe) - enum-int compare(GDScript)
- [Gut](https://github.com/bitwes/Gut) - editor_hint(TODO), null problem(GDScript), MARGIN_RIGHT(probably TODO), not dictionaty(Project/GDScript), funcref(TODO)
- [Lorien](https://github.com/mbrlabs/Lorien) - Constant Initializer(GDScript), enum compare(GDScript), get_datetime(TODO)
- [Jetpaca](https://github.com/KOBUGE-Games/jetpaca) - looks that Godot4 doesn't allow to use `String && bool`(GDScript), nill objects(GDScript)
- [Qarminer](https://github.com/qarmin/Qarminer) - Perfect conversion(some code needs to be enabled manually due availability more types in Godot 4 like Vector2i)
- [Pixelorama](https://github.com/Orama-Interactive/Pixelorama/) - broken project.godot(not sure) ,missing popup hide(Window not longer is child of popup - Project), enum int compare(GDScript), window_tite(TODO),  Vector2.INF(GDScript), extends BaseButton not works(GDScript), Tween(too complicated), constant initializer
- [Godot Demo projects](https://github.com/godotengine/godot-demo-projects) - get_simple_path(not sure), `to_lower` not found in StringName(GDScript)
- [Material Maker](https://github.com/RodZill4/material-maker) - Crash in `parse_dictionary`(GDScript)
- [Godot Next](https://github.com/godot-extended-libraries/godot-next) - Cannot add String to StringName(GDScript), Function "keys" has the same name as a previously declared variable(Project)
- [Tanks of Freedom 3D](https://github.com/P1X-in/Tanks-of-Freedom-3-D) - Cannot import(Godot)
- [TPS Demo](https://github.com/godotengine/tps-demo) - depth_draw_alpha_prepass shader(Godot?), await(TODO),
- [HeightMap plugin](https://github.com/Zylann/godot_heightmap_plugin) - already created variables(Project)

Legend:
- Project - project problem, not sure but may be probably added check to converter
- Godot - Godot bug(general)
- GDScript - GDScript bug
- TODO - I may this implement in future

**How to find functions that needs to be converted**
- The simplest(and hardest at once) method is to find manually each not working function and also find alternative function which works similarly or just is the same function but only with changed name
- More automated option require to:
  - compile godot from this PR
  - run this script with the latest available Godot 3 version - [PrintAllMethods3.zip](https://github.com/godotengine/godot/files/8504535/PrintAllMethods3.zip) (it will produce several txt files like `classes.txt` which contains info about almost all possible functions, classes, enums etc.)
  - run different script with the latest available Godot 4 version - [PrintAllMethods4.zip](https://github.com/godotengine/godot/files/8504534/PrintAllMethods4.zip)
  - copy produced files from both runs to new folder
  - create `project.godot` inside this folder
  - run command `godot4 --convert-to-godot40  --headless` (This will convert all things from `.gd` files)
  - sort results, remove duplicate lines etc.
```
sed -i 's/StringName/String/g' functions4.txt
sed -i 's/Vector2i/Vector2/g' functions4.txt
sed -i 's/Vector3i/Vector3/g' functions4.txt

sort classes4.txt | uniq > classes4.txt2;mv classes4.txt2 classes4.txt
sort constants4.txt | uniq > constants4.txt2;mv constants4.txt2 constants4.txt
sort enums4.txt | uniq > enums4.txt2;mv enums4.txt2 enums4.txt
sort functions4.txt | uniq > functions4.txt2;mv functions4.txt2 functions4.txt
sort functions_default4.txt | uniq > functions_default4.txt2;mv functions_default4.txt2 functions_default4.txt
sort properties4.txt | uniq > properties4.txt2;mv properties4.txt2 properties4.txt
sort signals4.txt | uniq > signals4.txt2;mv signals4.txt2 signals4.txt

sort classes3.gd | uniq > classes3.gd2;mv classes3.gd2 classes3.gd
sort constants3.gd | uniq > constants3.gd2;mv constants3.gd2 constants3.gd
sort enums3.gd | uniq > enums3.gd2;mv enums3.gd2 enums3.gd
sort functions3.gd | uniq > functions3.gd2;mv functions3.gd2 functions3.gd
sort functions_default3.gd | uniq > functions_default3.gd2;mv functions_default3.gd2 functions_default3.gd
sort properties3.gd | uniq > properties3.gd2;mv properties3.gd2 properties3.gd
sort signals3.gd | uniq > signals3.gd2;mv signals3.gd2 signals3.gd
```
  - compare `classes3.gd` with `classes4.txt` etc. with [Meld](https://meld.app/) `apt install meld`
  - it is visible now which classes/functions have same name or are already handled by converter
  
![Screenshot from 2022-04-03 15-32-00](https://user-images.githubusercontent.com/41945903/161430574-4340c0ac-c9a3-495a-9319-54444148a71c.png)

qarmin commented:
It works now, but I don't why it prints(CI fails due this)
```
[ 67%] editor/converter.cpp:383:13: warning: 'void converter()' defined but not used [-Wunused-function]
  383 | static void converter() {
      |             ^~~~~~~~~
```

lyuma commented:
Hey @qarmin 
I went ahead and merged in some of the extra function calls from my conversion scripts and committed them here:
https://github.com/godotengine/godot/pull/52093/commits/e5803a074c521afd00f8558e5864bc4fa27da821
A lot of them are for builtin types; others are renaming functions on classes that were also renamed. Whatever script was used to generate the original list probably missed those.

I think some of the ones that are commented out that are ambiguous are kind of important and probably better to assume the common case. For example, I consider Control.get_icon->get_theme_icon to be worth doing automatically, even though there are some rare cases in the Theme class and the TreeItem class that this would break.

FuncRef -> Callable is another one that needs to be addressed

There are many patterns that would be nice to do if we had a place to add more general purpose regular expressions that don't fit neatly into the functions or methods buckets:

Here are some regexes to handle cases I found to be common in my code, such as `Transform.xform` or `connect()`, using python's raw string syntax r"":
```py
	("\\.xform", "*"), # Transform3D
	(r"([\w.]+)\s*\.\s*xform_inv\s*([^)]*)", r"((\2) * (\1))"), # Transform3D

	("pause_mode = 2", "process_mode = "),
	("pause_mode", "process_mode"),
	(r"\b(connect|disconnect|is_connected)(\s*)\((\s*)([^,]*),(\s*)(?!Callable)([^,\s][^,]*),([^,]*)([,)])", r"\1\2(\3\4,\5Callable(\6,\7)\8"),
	(r"\byield\s*\((.*?)\)\s*$|\byield (.*)$", "await \\1\\2"),
	(r"\byield\s*\(\s*(\S[\s\S]*?)(\s*,\s*)*[\"'](.*)[\"']\s*\)\s*$", "await \\1.\\3"),
	(r"\b@?master(\s*)func\b", "@rpc(master)\\1func"),
	(r"\b@?puppet(\s*)func\b", "@rpc(puppet)\\1func"),
	(r"\b@?remote(\s*)func\b", "@rpc(any)\\1func"),
	(r"\b@?remotesync(\s*)func\b", "@rpc(any,sync)\\1func"),
	(r"\bfunc(\s+)_init(\s*)\(([^)]*)\)[^:]*:", r"func\1_init\2(\3):"),
	(r"\bget_node(\s+)\((\s+)\@", r"get_node\1(\2^"),
```
These replacements would be only for EditorPlugin: adding the missing _ to some of the overridable functions like get_name and so on.
```py
	(r"\bfunc(\s+)edit\b", r'func\1_edit', ['EditorPlugin']),
	(r"\bfunc(\s+)get_name\b", r'func\1_get_plugin_name', ['EditorPlugin']),
	(r"\bfunc(\s+)handles\b", r'func\1_handles', ['EditorPlugin']),
	(r"\bfunc(\s+)make_visible\b", r'func\1_make_visible', ['EditorPlugin']),
```

Finally some for `export`->`@export` syntax and `setget`->`: set=..., get=` syntax
```
EXPORT_PAREN = re.compile(r"^( *)(\bexport *)?(?:\(([^)]*)\))?( *var *[^=:]*)(: [^ =]*)?( *=[\s\S]*)?")
SETGET_PATTERN = re.compile(r"([\s\S]*)\bsetget\s*([^,]*)(?:,\s*)?([\s\S]*)")
```
[My python script here](https://github.com/V-Sekai/gd3to4/blob/main/godot4migrator.py) has a bunch of code to handle that. The export one breaks in a few cases, for example if the line contains a comma. Replacing setget seems moderately reliable.

Clearly I can understand if these heuristics and regexes wouldn't make it in, but I just want to get the discussion going on some other legitimately useful changes.

With all of my regexes as well as the function&class name replacements which your change does well, it usually only requires a few small edits and replaces to fix the rest. The biggest one that my regex usually messes up is mismatching parentheses on the connect syntax, when used like if (something.connect("signal", self, "method")==OK)

qarmin commented:
This PR should be created in my repository, because for now there is 2 different PR which implements a little different functions(I added new commit yesterday and I will add probably another today).
Changes to functions are mostly OK, but I didn't like mixing function names with different things like
```
{ "cursor_set_line", "set_caret_line" }, // Good, only renames function
{ "OS\\.get_ticks_msec", "Time.get_ticks_msec" }, // Bad, renames class but only for some functions 
``` 
Non casual renames like e.g. `(r"\bfunc(\s+)edit\b", r'func\1_edit', ['EditorPlugin']),` should be handled in `rename_enums`, `rename_classes` etc. functions(I don't like too much regex so it may took a while until I implement this).

qarmin commented:
@lyuma can you add real life examples(or even simplified) about your regex's from above/python script?
I'm quite bad at reading them, because a lot of better I work with this
```
(connect(A,B,C,D,E,F,G) != OK):  ->  (connect(A,Callable(B,C),D,E,F,G) != OK):
```
than 
```
RegEx("\\bconnect\\(([^,^\n]+),([^,^\n]+),([^,^\n^\\)]+)");
reg_connect.sub(file_content, "connect($1, Callable($2,$3)", true);
```

Calinou commented:
> Cyclic renames - e.g. in Skeleton3D choose_it change to select_this, but Node3D changes select_this to choose_it. So running tool, will renames all select_this functions to choose_it(assuming that this records have bigger index in function_renames array). Depends of the usage, some things may be renamed(this will broke other functions) when class is used more(like Node3D vs HTTPRequest) Cyclic renames are validated by tool.

This could be handled by renaming `choose_it` to `something_temporary`, renaming `select_this` to `choose_it` then renaming `something_temporary` to `select_this`.

qarmin commented:
Adding support for this, probably will require to add new array/function.
I'm not exactly sure if it is worth to do it - it will add some complexity but this problem probably happens only for some(I think max 10) methods.
I think it would be best to standardize these names, to not have such problems.

qarmin commented:
Finally I tested this on real project - Pixelorama, and looks that renamer works fine(even if output is polluted by hundreds of GDScript errors and needs several manual fixes).

For now I see only 1 font resource which is broken by tool, but when more issues with GDScript will be fixed, then more issues will be visible.

I can't test this more, because project freeze when parsing one GDScript file.

qarmin commented:
@aaronfranke @lyuma Do you have more examples what could be also replaced?
I've read your scripts and I think that I have implemented almost all things from them.

Now I'm waiting for merge #52571 and #52075 which will help to find more differences between versions

aaronfranke commented:
@qarmin Try using this tool on the Godot demo projects and see if there's anything else that can be automatically fixed. This is what I did to create my script, but the script is not a complete list of what can be automatically fixed in the demos.

Kemeros commented:
@qarmin Don't know if this is helpful or within scope but GDQuest has a repository with a few design pattern demos (state machine and ECS) in both GDScript and C#. He does some black magic(for someone of my level at least) with references and i am not able to manually convert his demos to GDScript 2.0. I was able to do part of it only.

Maybe this could give you ideas.

https://github.com/GDQuest/godot-design-patterns

KoBeWi commented:
I just noticed this in the OP

> await - not sure if yield -> await would easily convertable

Changing `yield(object, signal)` to `await object.signal` works in most cases. Only function yields can't be converted easily (you need to await whole stack).

> If want to have less amount of problems with converting, don't add Tweens to tree(try to instantitate them where needed).

If you use the [TweenSequence addon](https://github.com/godot-extended-libraries/godot-next/pull/50), it can be converted very easily.

btw, is scene conversion planned? There were some node properties or default values that have changed, also there are built-in resources (especially scripts) that might need conversion.

qarmin commented:
For now tool only search for files with gd, shader, tscn, tres, godot, cs and csproj extensions and for each type of file, execute different set of functions.

E.g. in gd files there is automatic conversion from `instance` -> `instantiate`, but not for `tres` or `tscn`, because it is used as keyword
```
[node name="HUD" parent="HUD" instance=ExtResource( 5 )]
```

fire commented:
If a file in the project is above 250KB ` ERROR: File has exceeded the maximum size allowed - 250 KB, file size - 266944 KB` it would not convert and leave the project half converted and in a broken state.

qarmin commented:
I added such limit, because some big files converted even in ~30 seconds, so at least testing is now faster and this limit can be changed to higher or even removed later.

Zireael07 commented:
Can the "too big file" error message at least mention the name of the file, please?

aaronfranke commented on editor/converter.cpp:
```diff
@@ -0,0 +1,3042 @@
+	{ "white", "WHITE" },
+	{ "whitesmoke", "WHITE_SMOKE" },
+	{ "yellow", "YELLOW" },
+	{ "yellowgreen", "YELLOW_GREEN " },
```
```suggestion
	{ "yellowgreen", "YELLOW_GREEN" },
```

qarmin replied:
This was intentional.
I want to see if CI will fail

Whimfoome commented:
Update: Amazing!
Here are some of the things I had to manually change while porting my project:
- `OS.GetWindowSafeArea()` -> `DisplayServer.ScreenGetUsableRect()` - many methods from OS were moved to DisplayServer
- `GetViewport().GetCamera()` -> `GetViewport().GetCamera3d()`
- In signals, I noticed that the `Connect()` arguments were automatically converted with `new Callable()`, thanks!, but the methods `IsConnected()` and `Disconnected()` follow the same approach too!
- Overriding `GetDragData` and `DropData` methods in the Control class now have a prefix `_`, so they should be renamed to `_GetDragData` and `_DropData`
- `BbcodeText` -> `Text`
- `Texture2D` -> `Texture`
- `Transform` -> `Transform3D`

qarmin commented:
I mainly use GDScript, so probably a lot of C# renames is missing
Things like `Texture` -> `Texture2D` are valid in GDScript because Texture is used only as class name(in C# probably this is used as property but probably can be used in  `SetTexture` way)

TokageItLab commented:
I recommend to make changes for the ping-pong loop.
```
has_loop = false -> loop_mode = 0 (Animation::LoopMode::LOOP_NONE)
has_loop = true -> loop_mode = 1 (Animation::LoopMode::LOOP_LINEAR)
```

EIREXE commented:
If you want I can give you access to the project heartbeat code for testing, it's a sizeable codebase so it might be a good test case

Zireael07 commented:
One thing I'm struggling with is get_global_transform().xform() and xform_inv() - I think most of my projects have this idiom somewhere and I can't figure out what the 4.0 equivalent is. Considering transforming stuff is a fairly basic operation, it would be worth it imho to have it on the list of auto-fixes...

raulsntos commented:
@Zireael07 the equivalent of `xform` and `xform_inv` in 4.0 is the `*` operator.

qarmin commented:
xform is already handled, but not sure if correctly, 
this are tests from PR
```
valid = valid & test_conversion_single_additional(" Transform.xform(Vector3(a,b,c)) ", " Transform * Vector3(a,b,c) ", &GodotConverter4::rename_gdscript_functions, "custom rename");

valid = valid & test_conversion_single_additional(" Transform.xform_inv(Vector3(a,b,c)) ", " Transform / Vector3(a,b,c) ", &GodotConverter4::rename_gdscript_functions, "custom rename"); // TODO, it is valid?
```

raulsntos commented:
@qarmin the equivalent of `xform_inv` is also the `*` operator just inverting the order, something like this:
```gdscript
# 3.0 = 4.0
Transform.xform(Vector3(a, b, c)) = Transform * Vector3(a, b, c)
Transform.xform_inv(Vector3(a, b, c)) = Vector3(a, b, c) * Transform
```

lyuma commented:
Good catch.
The diff in db18545cf113ba95bc5c7950c43649bbfae794b8 looks pretty good as far as converting A.xform_inv(B) to B * A

I've already (mostly) completed a port to 4.0, but when then engine is further along in alpha, I could run it on my 430-file 3.x codebase at some point and see what it gets right.

and-rad commented:
Should animation players be added to this? Since `transform` tracks got removed from the engine, none of these will be recognized by Godot 4, which leads to an empty animation player in the worst case scenario where all of the tracks are transform tracks. #55292 is such a case.

Replacing a single transform track with three individual loc/rot/scale tracks should be fairly straightforward, right?

qarmin commented:
Changing this is tres file is possible, but I need example how it looks in Godot 3 and how should it look in Godot 4 e.g.
```
anim/tracks = PoolByteArray(1 ,2, 3)
```
to 
```
anim/pos = PackedByteArray(3)
anim/rot = PackedByteArray(2)
anim/tra = PackedByteArray(1)
```

Since also binary files like `.anim` and `.res` are changed, I think that this should be fixed directly in animation importer, because this tool cannot modify such files

Zireael07 commented:
Bump. I have a project that I'd like to port to 4.0, and it's big enough that manually fixing all of the problems is untenable. Can this be merged in ASAP so that people like me, who have projects to port, can make use of this (and possibly submit improvement requests) instead of this languishing in the PR queue?

Whimfoome commented:
> Bump. I have a project that I'd like to port to 4.0, and it's big enough that manually fixing all of the problems is untenable. Can this be merged in ASAP so that people like me, who have projects to port, can make use of this (and possibly submit improvement requests) instead of this languishing in the PR queue?

I don't think it's ready for prime time yet. The only way to upgrade project with this PR right now is through the command line.
I'm not sure, but I think the goal is to make it seamless through the editor.
If you want to use the functionality right now, I would suggest just to build his PR with debug build, on my old 4 core CPU, it takes around 30 min.

qarmin commented:
At first https://github.com/godotengine/godot/pull/49053 needs to be merged, but still waits for reduz approval.
For now this PR is a little outdated, because some things changed and I don't rebase it too often.

I wait for clear message what exactly would happen with it, because if there is or will be another/better solution to upgrade Godot projects, then I don't want to waste time on updating this PR constantly.

Zireael07 commented:
Ping! Godot 4.0 alpha 1 just came out, and it's missing this incredibly vital feature!
I'd found some sort of a perl gist on GitHub, but it's missing quite a lot of stuff so I still have to do a lot by hand...

akien-mga commented:
> This PR contains #49053, so it would be good to merge that PR before this because it is needed to check if all functions exists even in builtin types.

Where is this method used exactly in this PR? I don't see any code that calls `ClassDB::get_variant_method_list`.

If it's not directly needed right now, it might be easier to merge the rest of the converter as agreeing on how to expose `get_variant_method_list` was the main point of debate.

qarmin committed:
Add Godot 3.x -> Godot 4.x project converter

qarmin commented:
`get_method_list_by_type` function was used to get list of methods from builtin types.

I expected that this code will find more breaking changes, but looks that not found any(at least yet), so it is not so important as I think that it would be.

I commented this code.

fire-forge commented on editor/converter.cpp:
```diff
@@ -0,0 +1,3667 @@
+	{ "physical_scancode", "physical_keycode" }, // InputEventKey
+	{ "popup_exclusive", "exclusive" }, // Window
+	{ "region_filter_clip", "region_filter_clip_enabled" }, // Sprite2D
+	{ "size_override_stretch", "size_2d_override_stretch" }, // SubViewport
```
Environment SSR properties were renamed in https://github.com/godotengine/godot/pull/59526

```suggestion
	{ "ss_reflections_enabled", "ssr_enabled" }, // Environment
	{ "ss_reflections_max_steps", "ssr_max_steps" }, // Environment
	{ "ss_reflections_fade_in", "ssr_fade_in" }, // Environment
	{ "ss_reflections_fade_out", "ssr_fade_out" }, // Environment
	{ "ss_reflections_depth_tolerance", "ssr_depth_tolerance" }, // Environment
```
Ignore the line removal in the suggestion, I can't figure out how to suggest a change without removing a line :/

duchainer commented:
Not certain this is the best place to put it, but:

If you want to add an Open Source project to test it on and compare with a real life conversion : I'm currently porting[ Librerama from Godot 3.4](https://codeberg.org/Librerama/librerama) [to Godot4.alpha7 in my own fork for now](https://codeberg.org/duchainer/librerama/src/branch/migrate-to-godot4-alpha7). 
The first commit of my new branch is the auto-converted stuff from the existing Godot3 to Godot4 converting taking place in alpha7. And the rest are the manual conversion I'm currently doing. 
I'll edit this post when I'm done, but I hope this can be useful.

Wish you the best, have a great weekend ^^

fire commented:
What are the blockers on merging this in a timely matter now?

fire-forge commented:
Does the converter support theme property renames? Several theme properties were renamed in https://github.com/godotengine/godot/pull/60261, and there may be others as well.

Calinou commented:
One more [shader rename](https://github.com/godotengine/godot/issues/61098) that could be added: `TRANSMISSION` → `SSS_TRANSMITTANCE_COLOR`

akien-mga commented:
I pushed an update to improve some names:
- `GodotConverter4` -> `ProjectConverter3To4`
- `converter.h` -> `project_converter_3_to_4.h`
- `--convert-to-godot40` -> `--convert-3to4`
- `--validate-convert-to-godot40` -> `--validation-conversion-3to4`

And fixed build with RegEx module disabled.

akien-mga reviewed (approved):
It's a pretty good start for converting projects. There are more conversions that can be added in future PRs, and still some manual work needed, but it does make it possible to open a fairly big project like Jetpaca in 4.0 and run it (with bugs) and with minor manual script changes.

akien-mga commented:
Thanks!

golddotasksquestions commented:
Is this the right place to submit remaining methods which still need to be included in the converter as of Beta12?
